File: action-type/14-stratery/go/pay_context.go
package main

import "math/big"

// PayContext 上下文，持有当前的支付策略
// 可以动态切换
type PayContext struct {
	strategy PayStrategy
}

// SetStrategy 设置当前支付策略
func (c *PayContext) SetStrategy(strategy PayStrategy) {

	c.strategy = strategy
}

// Pay 执行支付动作
func (c *PayContext) Pay(amount *big.Float) {

	if c.strategy == nil {
		panic("未设置支付策略")
	}

	c.strategy.Pay(amount)
}

File: action-type/14-stratery/java/PayContext.java
package main;

import java.math.BigDecimal;

public class PayContext {

    // ctx 持有的支付策略
    private PayStrategy strategy;

    public void setStrategy(PayStrategy strategy) {
        this.strategy = strategy;
    }

    // ctx 中的执行支付方法
    public void pay(BigDecimal amount) {
        
        if (strategy == null) {
            throw new IllegalStateException("未设置支付策略");
        }
        
        strategy.pay(amount);
    }
}
